### PR TITLE
[Query] fix fanout storage temporal panic

### PR DESCRIPTION
**Pull request by arnikola**
**What this PR does / why we need it**:
Fixes a panic caused when using temporal functions across multiple zones, adds tests.

**Special notes for your reviewer**:
The meat of this is refactoring how block Metadata is accessed (it was previously only accessible from block iterators, which would cause weird scenarios where iterators were generated only for access to metadata)

The actual bug fix is a 1-line change to array building for the unconsolidated block list in container.go#357

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
None

**Does this PR require updating code package or user-facing documentation?**:
None

### CODE REVIEW

File: src/query/block/container.go [hunk 1]
```diff
@@ -22,38 +22,71 @@ package block
...
-func newContainerBlock(blocks []Block) AccumulatorBlock {
+func newContainerBlock(blocks []Block) (AccumulatorBlock, error) {
+	if len(blocks) == 0 {
+		return nil, errNoContainerBlocks
+	}
+
+	meta := blocks[0].Meta()
+	for _, b := range blocks[1:] {
+		m := b.Meta()
+		// FIXME ARNIKOLA use actual equality here
```
Comment (benraskin92) [inline src/query/block/container.go]: ??

File: src/query/block/container.go [hunk 2]
```diff
@@ -22,38 +22,71 @@ package block
...
+func (b *containerBlock) Meta() Metadata {
+	return b.meta
+}
+
 func (b *containerBlock) AddBlock(bl Block) error {
 	if b.err != nil {
 		return b.err
 	}
 
+	// FIXME ARNIKOLA
```
Comment (benraskin92) [inline src/query/block/container.go]: ??

File: src/query/ts/m3db/encoded_block.go [hunk 3]
```diff
@@ -65,6 +65,7 @@ func NewEncodedBlock(
 		lastBlock,
 		opts,
 	)
+
 	err := bl.generateMetas()
```
Comment (benraskin92) [inline src/query/ts/m3db/encoded_block.go]: May as well make this:
```
if err := bl.generateMetas(); err != nil {
   return nil, err 
}
```

File: src/query/block/container.go [hunk 4]
```diff
@@ -312,40 +341,31 @@ func (it *containerSeriesIter) Current() Series {
...
+	blocks := make([]UnconsolidatedBlock, 0, len(b.blocks))
+	for _, bl := range b.blocks {
 		unconsolidated, err := bl.Unconsolidated()
 		if err != nil {
 			b.err = err
 			return nil, err
 		}
 
-		ucBlock.blocks[i] = unconsolidated
+		blocks = append(blocks, unconsolidated)
```
Comment (benraskin92) [inline src/query/block/container.go]: Nice

File: src/query/api/v1/handler/prometheus/native/read_common.go [hunk 5]
```diff
@@ -134,19 +134,25 @@ func sortedBlocksToSeriesList(blockList []blockWithMeta) ([]*ts.Series, error) {
 		return emptySeriesList, nil
 	}
 
-	firstBlock := blockList[0].block
+	var (
+		firstBlock = blockList[0].block
```
[the remark this answers is not in the document]
Comment (benraskin92) [reply src/query/api/v1/handler/prometheus/native/read_common.go]: Ignore - didn't see check above..

File: src/query/block/container.go [hunk 6]
```diff
@@ -22,38 +22,71 @@ package block
...
 }
 
-func newContainerBlock(blocks []Block) AccumulatorBlock {
+func newContainerBlock(blocks []Block) (AccumulatorBlock, error) {
+	if len(blocks) == 0 {
+		return nil, errNoContainerBlocks
+	}
+
+	meta := blocks[0].Meta()
+	for _, b := range blocks[1:] {
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/block/container.go]: Checked above :+1:

File: src/query/block/container.go (hunk 1 above)
Comment (arnikola) [reply src/query/block/container.go]: Bad comment, but waiting on https://github.com/m3db/m3/pull/1871 which adds an `Equals` function to the metas

File: src/query/block/container.go (hunk 2 above)
Comment (arnikola) [reply src/query/block/container.go]: Same here re: https://github.com/m3db/m3/pull/1871

File: src/query/api/v1/handler/prometheus/native/read_common.go [hunk 7]
```diff
@@ -134,19 +134,25 @@ func sortedBlocksToSeriesList(blockList []blockWithMeta) ([]*ts.Series, error) {
...
 
-	numSeries := firstSeriesIter.SeriesCount()
-	seriesMeta := firstSeriesIter.SeriesMeta()
-	bounds := firstSeriesIter.Meta().Bounds
-	commonTags := firstSeriesIter.Meta().Tags.Tags
+	var (
+		numSeries   = firstSeriesIter.SeriesCount()
+		seriesMeta  = firstSeriesIter.SeriesMeta()
+		seriesList  = make([]*ts.Series, numSeries)
+		seriesIters = make([]block.SeriesIter, len(blockList))
```
Comment (robskillington) [inline src/query/api/v1/handler/prometheus/native/read_common.go]: Maybe we should opt for using `make(thing, 0, expectedLength)` and using `append(..)` wherever possible?

Comment (arnikola) [reply src/query/api/v1/handler/prometheus/native/read_common.go]: Sounds good; I've tended to use indexed insertion when we know the slice length, but I guess for a fairly cold path it doesn't really matter and the safety is a bigger draw

File: src/query/block/container.go (hunk 4 above)
Comment (robskillington) [reply src/query/block/container.go]: As I mentioned prior, yeah maybe we should just be using `make(thing, 0, expectedLength)` and `append(...)` everywhere possible just to try to mainly avoid this being a thing in other places.

File: src/query/block/container_test.go [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+	assert.Equal(t, numSteps, it.StepCount())
+
+	assert.NoError(t, it.Err())
+	assert.NotPanics(t, func() { it.Close() })
+}
+
+//
+//
+//
+//
```
Comment (robskillington) [inline src/query/block/container_test.go]: Probably can remove these "//" before we merge?

File: src/query/api/v1/handler/prometheus/native/read_common.go (hunk 7 above)
Comment (arnikola) [reply src/query/api/v1/handler/prometheus/native/read_common.go]: (especially considering that's literally the root cause of this PR haha)

File: src/query/functions/temporal/base.go [hunk 9]
```diff
@@ -78,12 +81,9 @@ func (o baseOp) Node(controller *transform.Controller, opts transform.Options) t
 	}
 }
 
-// baseNode is an execution node
+// baseNode is an execution node.
 type baseNode struct {
-	op baseOp
-	// controller uses an interface here so we can mock it out in tests.
-	// TODO: use an exported interface everywhere instead of *transform.Controller.
-	// https://github.com/m3db/m3/issues/1430
```
Comment (robskillington) [inline src/query/functions/temporal/base.go]: Should we leave the note about #1430 or just close the issue if we don't intend to do this anymore?

File: src/query/block/container_test.go (hunk 8 above)
Comment (arnikola) [reply src/query/block/container_test.go]: RIP LOC stats 😭

File: src/query/storage/fanout/storage.go [hunk 10]
```diff
@@ -91,35 +93,60 @@ func (s *fanoutStorage) FetchBlocks(
...
-		}
+	var (
+		mu       sync.Mutex
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+	)
 
-		for _, bl := range result.Blocks {
-			it, err := bl.SeriesIter()
+	blockResult := make(map[string]block.Block, len(stores))
```
Comment (robskillington) [inline src/query/storage/fanout/storage.go]: Should we leave the comment about using a genny map?

File: src/query/functions/temporal/base.go (hunk 9 above)
Comment (arnikola) [reply src/query/functions/temporal/base.go]: Ah yeah my bad; we did turn one of these helper structs into an interface but it wasn't `transform.Controller`

File: src/query/storage/fanout/storage.go [hunk 11]
```diff
@@ -91,35 +93,60 @@ func (s *fanoutStorage) FetchBlocks(
...
+							mu.Lock()
+							multiErr = multiErr.Add(err)
+							mu.Unlock()
+							return
+						}
+					} else {
+						var err error
+						blockResult[key], err = block.NewContainerBlock(foundBlock, bl)
+						if err != nil {
+							mu.Lock()
```
Comment (robskillington) [inline src/query/storage/fanout/storage.go]: Hm, isn't this a double lock here (I see it above the for loop)

File: src/query/storage/fanout/storage.go [hunk 12]
```diff
@@ -91,35 +93,60 @@ func (s *fanoutStorage) FetchBlocks(
...
+			mu.Lock()
+			defer mu.Unlock()
+			for _, bl := range result.Blocks {
+				key := bl.Meta().String()
+				if foundBlock, found := blockResult[key]; found {
+					// this block exists. Check to see if it's already an appendable block.
+					if acc, ok := foundBlock.(block.AccumulatorBlock); ok {
+						// already an accumulator block, add current block.
+						if err := acc.AddBlock(bl); err != nil {
+							mu.Lock()
```
Comment (robskillington) [inline src/query/storage/fanout/storage.go]: Hm, isn't this a double lock here (I see it above the for loop)

File: src/query/storage/fanout/storage.go (hunk 11 above)
Comment (arnikola) [reply src/query/storage/fanout/storage.go]: Oops yes, nice catch

File: src/query/storage/fanout/storage.go [hunk 13]
```diff
@@ -91,35 +93,60 @@ func (s *fanoutStorage) FetchBlocks(
...
+						if err != nil {
+							mu.Lock()
+							multiErr = multiErr.Add(err)
+							mu.Unlock()
+							return
+						}
 					}
 				} else {
-					blockResult[key] = block.NewContainerBlock(foundBlock, bl)
+					blockResult[key] = bl
```
Comment (robskillington) [inline src/query/storage/fanout/storage.go]: Can we simplify the outdents in this for loop a little with continues?

```go
			for _, bl := range result.Blocks {
				key := bl.Meta().String()
				foundBlock, found := blockResult[key]
				if !found {
					blockResult[key] = bl
					continue
				}

				// This block exists. Check to see if it's already an appendable block.
				acc, ok := foundBlock.(block.AccumulatorBlock)
				if ok {
					// Already an accumulator block, add current block.
					err := acc.AddBlock(bl)
					if err != nil {
						multiErr = multiErr.Add(err)
						return
					}
					continue
				}

				var err error
				blockResult[key], err = block.NewContainerBlock(foundBlock, bl)
				if err != nil {
					multiErr = multiErr.Add(err)
					return
				}
			}


```

Comment (arnikola) [reply src/query/storage/fanout/storage.go]: Sounds good 👍

File: src/query/functions/temporal/base.go [hunk 14]
```diff
@@ -290,12 +321,13 @@ func (c *baseNode) processSingleRequest(request processRequest) (block.Block, er
...
+	var (
+		meta       = request.blk.Meta()
+		bounds     = meta.Bounds
+		seriesMeta = seriesIter.SeriesMeta()
+	)
 
-	// rename series to exclude their __name__ tag
-	// TODO: why do we do this?
+	// rename series to exclude their __name__ tag as part of function processing.
 	resultSeriesMeta := make([]block.SeriesMeta, len(seriesMeta))
```
Comment (robskillington) [inline src/query/functions/temporal/base.go]: Maybe use `append` here too?

File: src/query/block/container_test.go [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+package block
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/ts"
+	"github.com/stretchr/testify/require"
```
Comment (robskillington) [inline src/query/block/container_test.go]: nit: This should also be included in bottom imports?

File: src/query/storage/fanout/storage.go [hunk 16]
```diff
@@ -91,35 +94,68 @@ func (s *fanoutStorage) FetchBlocks(
...
 					}
-				} else {
-					blockResult[key] = block.NewContainerBlock(foundBlock, bl)
+
+					continue
+				}
+
+				accumulator, ok := foundBlock.(block.AccumulatorBlock)
+				if !ok {
+					multiErr = multiErr.Add(fmt.Errorf("container block has incorrect type"))
```
Comment (robskillington) [inline src/query/storage/fanout/storage.go]: Hm, shouldn't this be a `return` after adding the error to it?

Comment (arnikola) [reply src/query/storage/fanout/storage.go]: Oops, nice catch

**Pull request merged**